components/permissions/google_sign_in/google_sign_in_permission_util.cc

```diff
@@ -0,0 +1,202 @@
+constexpr char kFirebaseContentSettingsPattern[] =
+    "https://[*.]firebaseapp.com/__/auth*";
+// ContentSettingsPattern accepts [*.] as wildcard for subdomains, while
+// URLPattern takes *.
+constexpr char kFirebaseUrlPattern[] = "https://*.firebaseapp.com/__/auth*";
+
+bool IsGoogleAuthUrl(const GURL& gurl, const bool check_origin_only) {
+  static const std::vector<URLPattern> auth_login_patterns({
+      URLPattern(URLPattern::SCHEME_HTTPS, kGoogleAuthPattern),
+      URLPattern(URLPattern::SCHEME_HTTPS, kFirebaseUrlPattern),
+  });
+  return std::any_of(auth_login_patterns.begin(), auth_login_patterns.end(),
+                     [&gurl, check_origin_only](URLPattern pattern) {
+                       if (check_origin_only) {
+                         return pattern.MatchesSecurityOrigin(gurl);
+                       }
+                       return pattern.MatchesURL(gurl);
+                     });
+}
+}  // namespace
+
+bool IsGoogleAuthRelatedRequest(const GURL& request_url,
+                                const GURL& request_initiator_url) {
+  return IsGoogleAuthUrl(request_url, false) &&
+         !IsGoogleAuthUrl(request_initiator_url, true) &&
+         !net::registry_controlled_domains::SameDomainOrHost(
+             request_initiator_url,
+             GURL(URLPattern(URLPattern::SCHEME_HTTPS, kGoogleAuthPattern)
+                      .host()),
+             net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+}
+
+bool IsGoogleSignInFeatureEnabled() {
+  return base::FeatureList::IsEnabled(
+      permissions::features::kBraveGoogleSignInPermission);
+}
+
+bool IsGoogleSignInPrefEnabled(PrefService* prefs) {
+  return prefs->FindPreference(kGoogleLoginControlType) &&
+         prefs->GetBoolean(kGoogleLoginControlType);
+}
+
+blink::mojom::PermissionStatus GetCurrentGoogleSignInPermissionStatus(
+    content::PermissionControllerDelegate* permission_controller,
+    content::WebContents* contents,
+    const GURL& request_initiator_url) {
+  return permission_controller->GetPermissionStatusForOrigin(
+      blink::PermissionType::BRAVE_GOOGLE_SIGN_IN,
+      contents->GetPrimaryMainFrame(), request_initiator_url);
+}
+
+void Set3pCookieException(HostContentSettingsMap* content_settings,
+                          const ContentSettingsPattern& embedding_pattern,
+                          const ContentSetting& content_setting) {
+  const std::vector<std::string> auth_content_settings_patterns(
+      {kGoogleAuthPattern, kFirebaseContentSettingsPattern});
+  // embedding_pattern is the website that is embedding the google sign in
+  // auth pattern is auth.google.com or firebaseapp.com
+  for (const auto& auth_url_pattern : auth_content_settings_patterns) {
+    auto auth_pattern = ContentSettingsPattern::FromString(auth_url_pattern);
+    content_settings->SetContentSettingCustomScope(
+        auth_pattern, embedding_pattern, ContentSettingsType::BRAVE_COOKIES,
```

Comment:
why are we adding these rules to `BRAVE_COOKIES`? these rules will be disabled if Shields are down. Is this intended?
I believe we should put these rules to the usual `ContentSettingsType::COOKIES`.

pls see this:
https://github.com/brave/brave-core/blob/23a4475751f4a5ee42eb90518d1b02efb8ef2b0e/components/content_settings/core/browser/brave_content_settings_pref_provider.cc#L699

Reply:
I'm confused, I thought we imposed 3rd party cookie rule checking through BRAVE_COOKIES? Isn't the default value of COOKIES always true? In my browsertest, the initial value for 

```
  EXPECT_EQ(content_settings()->GetContentSetting(
                GURL(kAccountsGoogleUrl), embedding_url_,
                ContentSettingsType::COOKIES),
            ContentSetting::CONTENT_SETTING_DEFAULT);
```
fails, because the ContentSetting is actually ALLOW.

Reply:
Yes, It's correct, because 3p blocking controlled by content_settings::CookieSettings::ShouldBlockThirdPartyCookies().

Reply:
@boocmp I saw that but `content_settings::CookieSettings::ShouldBlockThirdPartyCookies()` one is the general policy, no? That doesn't check per-site 3p cookie policy. FWIW 

```
content_settings->SetContentSettingCustomScope(
        request_pattern, embedding_pattern, ContentSettingsType::COOKIES,
        content_setting);
```
gives the expected rule showing up in the cookie rules.

Reply:
GetContentSetting iterates through providers that support the specified content type. If you don't set any custom cookie rules for the 'url' then GetContentSetting won't find anything in the PrefProvider, but it will find value in the DefaultProvider. There is no default provider for BRAVE_COOKIES (and in your browser test you get CONTENT_SETTING_DEFAULT), but there is one for COOKIES, so content_settings()->GetContentSetting(...COOKIES...) will never return CONTENT_SETTING_DEFAULT, because in this case the default value from the DefaultProvider should be returned, and this value is equal to ALLOW (actually depends on general COOKIES settings).

Reply:
Yeah that makes sense, thanks @boocmp! It helped to go through the `brave_content_settings_pref_provider.cc`. I'm now using `cookie_settings()->GetCookieSetting` to verify 3p-ness.